File: Toggly.FeatureManagement.Python/toggly/src/toggly/context.py
"""Evaluation context for feature flag evaluation."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping


@dataclass
class TogglyEntityContext:
    """Canonical entity instance for ContextProperty evaluation."""

    kind: str
    key: str
    attributes: dict[str, Any] = field(default_factory=dict)

    def get_attribute(self, name: str) -> Any:
        """Return an attribute by name, matching case-insensitively."""
        if name in self.attributes:
            return self.attributes[name]
        lower = name.lower()
        for key, value in self.attributes.items():
            if key.lower() == lower:
                return value
        return None

    def contains_attribute(self, name: str) -> bool:
        """Return whether an attribute exists, matching case-insensitively."""
        if name in self.attributes:
            return True
        lower = name.lower()
        return any(key.lower() == lower for key in self.attributes)


@dataclass
class RequestContext:
    """HTTP request fields used by segment identity filters."""

    user_agent: str | None = None
    """User-Agent header value."""

    accept_language: str | None = None
    """Accept-Language header value."""

    country: str | None = None
    """Country code (e.g. from CF-IPCountry)."""

    def to_dict(self) -> dict[str, Any]:
        """Convert to a dictionary (camelCase keys for EvalContext parity)."""
        return {
            "userAgent": self.user_agent,
            "acceptLanguage": self.accept_language,
            "country": self.country,
        }

    @classmethod
    def from_dict(cls, data: Mapping[str, Any] | None) -> RequestContext | None:
        """Create from a dictionary accepting camelCase or snake_case keys."""
        if not data:
            return None
        return cls(
            user_agent=_first_str(data, "userAgent", "user_agent"),
            accept_language=_first_str(data, "acceptLanguage", "accept_language"),
            country=_first_str(data, "country"),
        )


def _first_str(data: Mapping[str, Any], *keys: str) -> str | None:
    for key in keys:
        if key in data and data[key] is not None:
            value = str(data[key])
            return value if value else None
    return None


class HttpRequestMapper:
    """Maps common HTTP headers into RequestContext fields.

    Does not invent identity, groups, or claims — merge those separately.
    """

    @staticmethod
    def from_http_headers(headers: Mapping[str, str] | None) -> RequestContext:
        """Build RequestContext from a header bag (case-insensitive keys)."""
        if not headers:
            return RequestContext()
        return RequestContext(
            user_agent=_header(headers, "user-agent"),
            accept_language=_header(headers, "accept-language"),
            country=_first_present(
                headers,
                "cf-ipcountry",
                "x-vercel-ip-country",
                "cloudfront-viewer-country",
            ),
        )

    @staticmethod
    def merge_into(
        headers: Mapping[str, str] | None,
        base: EvaluationContext | None,
    ) -> EvaluationContext:
        """Merge HTTP-mapped request fields over an existing evaluation context."""
        request = HttpRequestMapper.from_http_headers(headers)
        if base is None:
            return EvaluationContext(request=request)
        return base.with_request(request)


def _header(headers: Mapping[str, str], name: str) -> str | None:
    lower = name.lower()
    for key, value in headers.items():
        if key is not None and key.lower() == lower and value:
            return value
    return None


def _first_present(headers: Mapping[str, str], *names: str) -> str | None:
    for name in names:
        value = _header(headers, name)
        if value is not None:
            return value
    return None


@dataclass
class EvaluationContext:
    """Context for evaluating feature flags.

    The evaluation context provides information about the current user,
    their groups, claims, request fields, and any additional attributes
    used for targeting and percentage rollouts.

    Example:
        >>> context = EvaluationContext(
        ...     identity="user-123",
        ...     groups=["beta-testers", "premium"],
        ...     claims={"role": "admin"},
        ...     request=RequestContext(country="US"),
        ...     traits={"plan": "enterprise"}
        ... )

    """

    identity: str | None = None
    """Unique identifier for the user/device. Used for deterministic rollouts."""

    groups: list[str] = field(default_factory=list)
    """Groups the user belongs to (e.g., 'beta-testers', 'admins')."""

    traits: dict[str, Any] = field(default_factory=dict)
    """Additional attributes for targeting (e.g., country, plan, version)."""

    claims: dict[str, str] = field(default_factory=dict)
    """Principal / JWT-style claims for UserClaims filters."""

    request: RequestContext | None = None
    """HTTP request fields for segment filters."""

    entity: TogglyEntityContext | None = None
    """Optional entity for ContextProperty filters."""

    def with_identity(self, identity: str) -> EvaluationContext:
        """Create a new context with the specified identity."""
        return EvaluationContext(
            identity=identity,
            groups=self.groups.copy(),
            traits=self.traits.copy(),
            claims=self.claims.copy(),
            request=self.request,
            entity=self.entity,
        )

    def with_groups(self, *groups: str) -> EvaluationContext:
        """Create a new context with additional groups."""
        return EvaluationContext(
            identity=self.identity,
            groups=list(set(self.groups + list(groups))),
            traits=self.traits.copy(),
            claims=self.claims.copy(),
            request=self.request,
            entity=self.entity,
        )

    def with_traits(self, **traits: Any) -> EvaluationContext:
        """Create a new context with additional traits."""
        merged_traits = {**self.traits, **traits}
        return EvaluationContext(
            identity=self.identity,
            groups=self.groups.copy(),
            traits=merged_traits,
            claims=self.claims.copy(),
            request=self.request,
            entity=self.entity,
        )

    def with_claims(self, claims: dict[str, str]) -> EvaluationContext:
        """Create a new context with the specified claims map."""
        return EvaluationContext(
            identity=self.identity,
            groups=self.groups.copy(),
            traits=self.traits.copy(),
            claims=dict(claims) if claims else {},
            request=self.request,
            entity=self.entity,
        )

    def with_request(self, request: RequestContext | None) -> EvaluationContext:
        """Create a new context with the specified request fields."""
        return EvaluationContext(
            identity=self.identity,
            groups=self.groups.copy(),
            traits=self.traits.copy(),
            claims=self.claims.copy(),
            request=request,
            entity=self.entity,
        )

    def with_entity(self, entity: TogglyEntityContext | None) -> EvaluationContext:
        """Create a new context with the specified entity."""
        return EvaluationContext(
            identity=self.identity,
            groups=self.groups.copy(),
            traits=self.traits.copy(),
            claims=self.claims.copy(),
            request=self.request,
            entity=entity,
        )

    def to_dict(self) -> dict[str, Any]:
        """Convert context to a dictionary."""
        return {
            "identity": self.identity,
            "groups": self.groups,
            "traits": self.traits,
            "claims": self.claims,
            "request": None if self.request is None else self.request.to_dict(),
            "entity": None
            if self.entity is None
            else {
                "kind": self.entity.kind,
                "key": self.entity.key,
                "attributes": self.entity.attributes,
            },
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> EvaluationContext:
        """Create a context from a dictionary."""
        entity_data = data.get("entity")
        entity = None
        if isinstance(entity_data, dict):
            entity = TogglyEntityContext(
                kind=str(entity_data.get("kind", "")),
                key=str(entity_data.get("key", "")),
                attributes=entity_data.get("attributes") or {},
            )
        claims_raw = data.get("claims") or {}
        claims = {str(k): str(v) for k, v in claims_raw.items()} if isinstance(claims_raw, dict) else {}
        request_data = data.get("request")
        request = RequestContext.from_dict(request_data) if isinstance(request_data, dict) else None
        return cls(
            identity=data.get("identity"),
            groups=data.get("groups", []),
            traits=data.get("traits", {}),
            claims=claims,
            request=request,
            entity=entity,
        )

    @classmethod
    def anonymous(cls) -> EvaluationContext:
        """Create an anonymous evaluation context."""
        return cls()

    def __bool__(self) -> bool:
        """Check if context has any meaningful data."""
        return bool(
            self.identity
            or self.groups
            or self.traits
            or self.claims
            or self.request
            or self.entity
        )

File: toggly-rust/toggly/src/context.rs
//! Evaluation context for feature flag checks.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Canonical entity instance for ContextProperty evaluation.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TogglyEntityContext {
    /// Context kind (Order, Product, etc.).
    pub kind: String,
    /// Entity key.
    pub key: String,
    /// Attribute map used by ContextProperty filters.
    pub attributes: HashMap<String, serde_json::Value>,
}

impl TogglyEntityContext {
    /// Look up an attribute, ignoring key case.
    pub fn get_attr(&self, name: &str) -> Option<&serde_json::Value> {
        if let Some(v) = self.attributes.get(name) {
            return Some(v);
        }
        self.attributes
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case(name))
            .map(|(_, v)| v)
    }

    /// Whether the attribute exists (case-insensitive).
    pub fn contains_attr(&self, name: &str) -> bool {
        self.attributes.contains_key(name)
            || self.attributes.keys().any(|k| k.eq_ignore_ascii_case(name))
    }
}

/// HTTP request fields used by segment identity filters.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct RequestContext {
    /// User-Agent header value.
    #[serde(default, alias = "user_agent", rename = "userAgent")]
    pub user_agent: Option<String>,

    /// Accept-Language header value.
    #[serde(default, alias = "accept_language", rename = "acceptLanguage")]
    pub accept_language: Option<String>,

    /// Country code (e.g. from CF-IPCountry).
    #[serde(default)]
    pub country: Option<String>,
}

impl RequestContext {
    /// Whether any request field is set.
    pub fn is_empty(&self) -> bool {
        self.user_agent.as_ref().is_none_or(|s| s.is_empty())
            && self.accept_language.as_ref().is_none_or(|s| s.is_empty())
            && self.country.as_ref().is_none_or(|s| s.is_empty())
    }
}

/// Maps common HTTP headers into [`RequestContext`] fields.
///
/// Does not invent identity, groups, or claims — merge those separately.
pub struct HttpRequestMapper;

impl HttpRequestMapper {
    /// Build [`RequestContext`] from a header bag (case-insensitive keys).
    pub fn from_http_headers<I, K, V>(headers: I) -> RequestContext
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: AsRef<str>,
    {
        let mut user_agent = None;
        let mut accept_language = None;
        let mut cf_country = None;
        let mut vercel_country = None;
        let mut cloudfront_country = None;

        for (key, value) in headers {
            let value = value.as_ref();
            if value.is_empty() {
                continue;
            }
            let lower = key.as_ref().to_ascii_lowercase();
            match lower.as_str() {
                "user-agent" if user_agent.is_none() => {
                    user_agent = Some(value.to_string());
                }
                "accept-language" if accept_language.is_none() => {
                    accept_language = Some(value.to_string());
                }
                "cf-ipcountry" if cf_country.is_none() => {
                    cf_country = Some(value.to_string());
                }
                "x-vercel-ip-country" if vercel_country.is_none() => {
                    vercel_country = Some(value.to_string());
                }
                "cloudfront-viewer-country" if cloudfront_country.is_none() => {
                    cloudfront_country = Some(value.to_string());
                }
                _ => {}
            }
        }

        RequestContext {
            user_agent,
            accept_language,
            country: cf_country.or(vercel_country).or(cloudfront_country),
        }
    }

    /// Merge HTTP-mapped request fields over an existing evaluation context.
    pub fn merge_into<I, K, V>(headers: I, base: Option<&EvalContext>) -> EvalContext
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: AsRef<str>,
    {
        let request = Self::from_http_headers(headers);
        match base {
            Some(base) => {
                let mut ctx = base.clone();
                ctx.request = Some(request);
                ctx
            }
            None => EvalContext {
                request: Some(request),
                ..Default::default()
            },
        }
    }
}

/// Context for feature flag evaluation.
///
/// Provides user information and attributes for targeting rules.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct EvalContext {
    /// Unique identifier for the user/device.
    pub identity: Option<String>,

    /// Group identifiers (e.g., "beta-testers", "premium").
    #[serde(default)]
    pub groups: Vec<String>,

    /// Principal / JWT-style claims for UserClaims filters.
    #[serde(default)]
    pub claims: HashMap<String, String>,

    /// Custom attributes for targeting rules.
    #[serde(default)]
    pub traits: HashMap<String, serde_json::Value>,

    /// HTTP request fields for segment filters.
    #[serde(default)]
    pub request: Option<RequestContext>,

    /// Entity for ContextProperty filters.
    #[serde(default)]
    pub entity: Option<TogglyEntityContext>,
}

impl EvalContext {
    /// Create a new evaluation context builder.
    pub fn builder() -> EvalContextBuilder {
        EvalContextBuilder::default()
    }

    /// Create a context with just an identity.
    pub fn with_identity(identity: impl Into<String>) -> Self {
        Self {
            identity: Some(identity.into()),
            ..Default::default()
        }
    }

    /// Check if the context has an identity.
    pub fn has_identity(&self) -> bool {
        self.identity.as_ref().is_some_and(|id| !id.is_empty())
    }

    /// Get a trait value.
    pub fn get_trait(&self, key: &str) -> Option<&serde_json::Value> {
        self.traits.get(key)
    }

    /// Check if the context belongs to a group.
    pub fn in_group(&self, group: &str) -> bool {
        self.groups.iter().any(|g| g == group)
    }

    /// Look up a claim by exact type key.
    pub fn get_claim(&self, claim_type: &str) -> Option<&str> {
        self.claims.get(claim_type).map(|s| s.as_str())
    }
}

/// Builder for [`EvalContext`].
#[derive(Debug, Default)]
pub struct EvalContextBuilder {
    context: EvalContext,
}

impl EvalContextBuilder {
    /// Set the identity.
    pub fn identity(mut self, identity: impl Into<String>) -> Self {
        self.context.identity = Some(identity.into());
        self
    }

    /// Set the groups.
    pub fn groups(mut self, groups: Vec<String>) -> Self {
        self.context.groups = groups;
        self
    }

    /// Add a group.
    pub fn group(mut self, group: impl Into<String>) -> Self {
        self.context.groups.push(group.into());
        self
    }

    /// Set principal claims.
    pub fn claims(mut self, claims: HashMap<String, String>) -> Self {
        self.context.claims = claims;
        self
    }

    /// Add a claim.
    pub fn claim(mut self, claim_type: impl Into<String>, value: impl Into<String>) -> Self {
        self.context
            .claims
            .insert(claim_type.into(), value.into());
        self
    }

    /// Set the traits.
    pub fn traits(mut self, traits: HashMap<String, serde_json::Value>) -> Self {
        self.context.traits = traits;
        self
    }

    /// Add a trait.
    pub fn trait_value(
        mut self,
        key: impl Into<String>,
        value: impl Into<serde_json::Value>,
    ) -> Self {
        self.context.traits.insert(key.into(), value.into());
        self
    }

    /// Set request fields for segment filters.
    pub fn request(mut self, request: RequestContext) -> Self {
        self.context.request = Some(request);
        self
    }

    /// Attach an entity for ContextProperty evaluation.
    pub fn entity(mut self, entity: TogglyEntityContext) -> Self {
        self.context.entity = Some(entity);
        self
    }

    /// Build the context.
    pub fn build(self) -> EvalContext {
        self.context
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_context() {
        let ctx = EvalContext::default();
        assert!(ctx.identity.is_none());
        assert!(ctx.groups.is_empty());
        assert!(ctx.claims.is_empty());
        assert!(ctx.traits.is_empty());
        assert!(ctx.request.is_none());
    }

    #[test]
    fn test_with_identity() {
        let ctx = EvalContext::with_identity("user-123");
        assert_eq!(ctx.identity, Some("user-123".to_string()));
    }

    #[test]
    fn test_builder() {
        let ctx = EvalContext::builder()
            .identity("user-123")
            .group("beta")
            .group("premium")
            .claim("role", "admin")
            .trait_value("country", "US")
            .trait_value("age", 25)
            .request(RequestContext {
                user_agent: Some("ua".into()),
                accept_language: Some("en-US".into()),
                country: Some("US".into()),
            })
            .build();

        assert_eq!(ctx.identity, Some("user-123".to_string()));
        assert_eq!(ctx.groups, vec!["beta", "premium"]);
        assert_eq!(ctx.get_claim("role"), Some("admin"));
        assert_eq!(ctx.get_trait("country"), Some(&serde_json::json!("US")));
        assert_eq!(ctx.get_trait("age"), Some(&serde_json::json!(25)));
        assert_eq!(
            ctx.request.as_ref().and_then(|r| r.country.as_deref()),
            Some("US")
        );
    }

    #[test]
    fn test_has_identity() {
        let ctx = EvalContext::default();
        assert!(!ctx.has_identity());

        let ctx = EvalContext::with_identity("");
        assert!(!ctx.has_identity());

        let ctx = EvalContext::with_identity("user-123");
        assert!(ctx.has_identity());
    }

    #[test]
    fn test_in_group() {
        let ctx = EvalContext::builder()
            .group("beta")
            .group("premium")
            .build();

        assert!(ctx.in_group("beta"));
        assert!(ctx.in_group("premium"));
        assert!(!ctx.in_group("admin"));
    }

    #[test]
    fn test_http_request_mapper_country_priority() {
        let mapped = HttpRequestMapper::from_http_headers([
            ("user-agent", "ua"),
            ("accept-language", "en"),
            ("cf-ipcountry", "US"),
            ("x-vercel-ip-country", "DE"),
            ("cloudfront-viewer-country", "FR"),
        ]);
        assert_eq!(mapped.user_agent.as_deref(), Some("ua"));
        assert_eq!(mapped.accept_language.as_deref(), Some("en"));
        assert_eq!(mapped.country.as_deref(), Some("US"));

        let vercel = HttpRequestMapper::from_http_headers([("x-vercel-ip-country", "DE")]);
        assert_eq!(vercel.country.as_deref(), Some("DE"));

        let cf = HttpRequestMapper::from_http_headers([("CloudFront-Viewer-Country", "FR")]);
        assert_eq!(cf.country.as_deref(), Some("FR"));
    }

    #[test]
    fn test_http_request_mapper_merge_into() {
        let base = EvalContext::builder().identity("u").group("beta").build();
        let merged = HttpRequestMapper::merge_into([("cf-ipcountry", "US")], Some(&base));
        assert_eq!(merged.identity.as_deref(), Some("u"));
        assert_eq!(merged.groups, vec!["beta"]);
        assert_eq!(
            merged.request.as_ref().and_then(|r| r.country.as_deref()),
            Some("US")
        );
    }

    #[test]
    fn test_request_serde_aliases() {
        let json = r#"{"userAgent":"ua","acceptLanguage":"en","country":"US"}"#;
        let request: RequestContext = serde_json::from_str(json).unwrap();
        assert_eq!(request.user_agent.as_deref(), Some("ua"));
        assert_eq!(request.accept_language.as_deref(), Some("en"));
        assert_eq!(request.country.as_deref(), Some("US"));
    }
}
